conda/__init__.py

```diff
@@ -148,3 +139,209 @@ def _default(self, obj):
+        remove_version = self._parse_version(remove_in)
+
+        addendum = f" {addendum}" if addendum else ""
+        if self._version < deprecate_version:
+            self._category = PendingDeprecationWarning
+            message = f"{{name}} is pending deprecation and will be removed in {remove_in}."
+        elif self._version < remove_version:
+            self._category = DeprecationWarning
+            message = f"{{name}} is deprecated and will be removed in {remove_in}."
+        else:
+            self._category = None
+            message = f"{{name}} was slated for removal in {remove_in}."
+        self._message = lambda name: f"{message}{addendum}".format(name=name)
+
+        self._stack = stack
+
+    def __call__(self, func: Callable) -> Callable:
+        """Deprecation decorator for functions, methods, & classes."""
+        # detect function name
+        fullname = f"{func.__module__}.{func.__qualname__}"
+        if self._argument:
+            fullname = f"{func.__module__}.{func.__qualname__}({self._argument})"
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(fullname))
+
+        # alert user that it's time to remove something
+        @wraps(func)
+        def inner(*args, **kwargs):
+            # always warn if not deprecating an argument
+            # only warn about argument deprecations if the argument is used
+            if not self._argument or self._argument in kwargs:
+                warnings.warn(self._message(fullname), self._category, stacklevel=2 + self._stack)
+
+                # rename argument deprecations as needed
+                value = kwargs.pop(self._argument, None)
+                if self._rename:
+                    kwargs.setdefault(self._rename, value)
+
+            return func(*args, **kwargs)
+
+        return inner
+
+    @classmethod
+    def argument(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        argument: str,
+        *,
+        rename: str | None = None,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation decorator for keyword arguments."""
+        # provide a default addendum if renaming and no addendum is provided
+        if rename and not addendum:
+            addendum = f"Use '{rename}' instead."
+
+        self = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)
+        self._argument = argument
+        self._rename = rename
+
+        return self
+
+    @staticmethod
+    def _get_module(stack: int) -> tuple[ModuleType, str]:
+        import inspect  # expensive
+
+        try:
+            frame = inspect.stack()[2 + stack]
+            module = inspect.getmodule(frame[0])
+            return (module, module.__name__)
+        except (IndexError, AttributeError):
+            raise RuntimeError("unable to determine the calling module") from None
```

Comment:
I'm explicitly catching any issues with the user-provided `stack` value used in `inspect.stack()[2 + stack]` since there is no guarantee that the user provided a reasonable (or even positive) value (e.g., `inf` or `-inf`).

Reply:
Oh, good point!